distributed/tests/test_coassignmnet_group.py

```diff
@@ -0,0 +1,298 @@
+    for values in cogroups.values():
+        assert not (seen & values)
+        seen.update(values)
+
+
+@pytest.fixture(params=["abcde"])
+def abcde(request):
+    return request.param
+
+
+def dummy_dsk_to_taskstate(dsk: dict) -> list[TaskState]:
+    task_groups: dict[str, TaskGroup] = {}
+    tasks = dict()
+    priority = dask.order.order(dsk)
+    for key in dsk:
+        tasks[key] = ts = TaskState(key, None, "released")
+        ts.group = task_groups.get(ts.group_key, TaskGroup(ts.group_key))
+        task_groups[ts.group_key] = ts.group
+        ts.priority = priority[key]
+    for key, vals in dsk.items():
+        stack = list(vals[1:])
+        while stack:
+            d = stack.pop()
+            if isinstance(d, list):
+                stack.extend(d)
+                continue
+            assert isinstance(d, (str, tuple, int))
+            if d not in tasks:
+                raise ValueError(f"Malformed example. {d} not part of dsk")
+            tasks[key].add_dependency(tasks[d])
+    return sorted(tasks.values(), key=lambda ts: ts.priority)
+
+
+def test_tree_reduce(abcde):
+    r""" """
+    a, b, c, _, _ = abcde
+    a1, a2, a3, a4, a5, a6, a7, a8, a9 = (a + i for i in "123456789")
+    b1, b2, b3, b4 = (b + i for i in "1234")
+    dsk = {
+        a1: (f,),
+        a2: (f,),
+        a3: (f,),
+        b1: (f, a1, a2, a3),
+        a4: (f,),
+        a5: (f,),
+        a6: (f,),
+        b2: (f, a4, a5, a6),
+        a7: (f,),
+        a8: (f,),
+        a9: (f,),
+        b3: (f, a7, a8, a9),
+        c: (f, b1, b2, b3),
+    }
+    tasks = dummy_dsk_to_taskstate(dsk)
+    assert isinstance(tasks, list)
+    cogroups = coassignmnet_groups(tasks)
+    assert_disjoint_sets(cogroups)
+    assert len(cogroups) == 3
+
+
+def test_nearest_neighbor(abcde):
+    r"""
+    a1  a2  a3  a4  a5  a6  a7 a8  a9
+     \  |  /  \ |  /  \ |  / \ |  /
+        b1      b2      b3     b4
+
+    No co-groups
+    """
+    a, b, c, _, _ = abcde
+    a1, a2, a3, a4, a5, a6, a7, a8, a9 = (a + i for i in "123456789")
+    b1, b2, b3, b4 = (b + i for i in "1234")
+
+    dsk = {
+        b1: (f,),
+        b2: (f,),
+        b3: (f,),
+        b4: (f,),
+        a1: (f, b1),
+        a2: (f, b1),
+        a3: (f, b1, b2),
+        a4: (f, b2),
+        a5: (f, b2, b3),
+        a6: (f, b3),
+        a7: (f, b3, b4),
+        a8: (f, b4),
+        a9: (f, b4),
+    }
+    tasks = dummy_dsk_to_taskstate(dsk)
+    assert isinstance(tasks, list)
+    cogroups = coassignmnet_groups(tasks)
+    assert_disjoint_sets(cogroups)
+    assert len(cogroups) == 0
+
+
+def test_deep_bases_win_over_dependents(abcde):
+    r"""
+    It's not clear who should run first, e or d
+
+    1.  d is nicer because it exposes parallelism
+    2.  e is nicer (hypothetically) because it will be sooner released
+        (though in this case we need d to run first regardless)
+
+    Regardless of e or d first, we should run b before c.
+
+            a
+          / | \   .
```

Comment:
```suggestion
          / | \
```

Reply:
FYI I literally just copied this stuff from dask.dask